### PR TITLE
Improve Arm Memory Attribute Protocol Logging

### LINKED ISSUE #124: [Bug]: Code marked as NX when boot Ubuntu
### Is there an existing issue for this?

- [X] I have searched existing issues

### Current Behavior

There was a report recently on booting Linux ARM64 system, where the system will run into synchronous exception after handing off to Linux bootloader. Where the exception occurs is at the boundary of a memory page. The debugger shows that the code to be executed is marked as not executable from MMU.

The OS is downloaded from below:
https://cdimages.ubuntu.com/jammy/daily-live/pending/

The behavior started to occur after picking up https://github.com/microsoft/mu_silicon_arm_tiano/commit/9bb8982dbc94d7e5d7a553ba80777e2c24f26328 from https://github.com/microsoft/mu_silicon_arm_tiano/commit/17f063737d448935f12696427c1ae6727995b0e8.

### Expected Behavior

System should boot properly.

### Steps To Reproduce

To be tested on QEMU. But otherwise, just build the aforementioned commit and try to boot Ubuntu system.

### Build Environment

```markdown
- OS(s): Ubuntu 2204 nightly build.
- Tool Chain(s): GCC5
- Targets Impacted: DEBUG,RELEASE
```

### Version Information

```text
https://github.com/microsoft/mu_silicon_arm_tiano/commit/9bb8982dbc94d7e5d7a553ba80777e2c24f26328
```

### Urgency

Medium

### Are you going to fix this?

Someone else needs to fix it

### Do you need maintainer feedback?

No maintainer feedback needed

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by os-d**
# Preface

Please ensure you have read the [contribution docs](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md) prior
to submitting the pull request. In particular,
[pull request guidelines](https://github.com/microsoft/mu/blob/master/CONTRIBUTING.md#pull-request-best-practices).

## Description

The Arm Memory Attribute Protocol logging is lacking, as evidenced by issue #124. This PR simply adds additional logging to make future debugging work easier and to help identify issue #124 in the wild, as it is likely to come up again with non-patched versions of the Linux shim.

The x86 Memory Attribute Protocol code already has good logging in the error cases.

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [ ] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

Only logging update, no testing.

## Integration Instructions

N/A.

### CODE REVIEW

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c [hunk 1]
```diff
@@ -84,6 +84,11 @@ GetMemoryAttributes (
   EFI_STATUS  Status;
 
   if ((Length == 0) || (Attributes == NULL)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a invalid length is 0 or attributes are null\n",
```
Comment (kuqin12) [inline ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: might want to print the length and pointer value here?

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c [hunk 2]
```diff
@@ -213,20 +225,41 @@ SetMemoryAttributes (
   if ((Attributes & EFI_MEMORY_RP) != 0) {
     Status = ArmSetMemoryRegionNoAccess (BaseAddress, Length);
     if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a failed to set EFI_MEMORY_RP on BaseAddress 0x%llx of length 0x%llx!\n",
```
Comment (kuqin12) [inline ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: also the returned status?

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c [hunk 3]
```diff
@@ -213,20 +225,41 @@ SetMemoryAttributes (
...
       return EFI_UNSUPPORTED;
     }
   }
 
   if ((Attributes & EFI_MEMORY_RO) != 0) {
     Status = ArmSetMemoryRegionReadOnly (BaseAddress, Length);
     if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a failed to set EFI_MEMORY_RO on BaseAddress 0x%llx of length 0x%llx!\n",
```
Comment (kuqin12) [inline ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Same here.

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c [hunk 4]
```diff
@@ -213,20 +225,41 @@ SetMemoryAttributes (
...
       return EFI_UNSUPPORTED;
     }
   }
 
   if ((Attributes & EFI_MEMORY_XP) != 0) {
     Status = ArmSetMemoryRegionNoExec (BaseAddress, Length);
     if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a failed to set EFI_MEMORY_XP on BaseAddress 0x%llx of length 0x%llx!\n",
```
Comment (kuqin12) [inline ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Ditto.

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c [hunk 5]
```diff
@@ -299,20 +339,41 @@ ClearMemoryAttributes (
   if ((Attributes & EFI_MEMORY_RP) != 0) {
     Status = ArmClearMemoryRegionNoAccess (BaseAddress, Length);
     if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a failed to clear EFI_MEMORY_RP from BaseAddress 0x%llx of length 0x%llx!\n",
```
Comment (kuqin12) [inline ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Ditto

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c [hunk 6]
```diff
@@ -299,20 +339,41 @@ ClearMemoryAttributes (
...
       return EFI_UNSUPPORTED;
     }
   }
 
   if ((Attributes & EFI_MEMORY_RO) != 0) {
     Status = ArmClearMemoryRegionReadOnly (BaseAddress, Length);
     if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a failed to clear EFI_MEMORY_RO from BaseAddress 0x%llx of length 0x%llx!\n",
```
Comment (kuqin12) [inline ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Ditto

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c [hunk 7]
```diff
@@ -299,20 +339,41 @@ ClearMemoryAttributes (
...
       return EFI_UNSUPPORTED;
     }
   }
 
   if ((Attributes & EFI_MEMORY_XP) != 0) {
     Status = ArmClearMemoryRegionNoExec (BaseAddress, Length);
     if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a failed to clear EFI_MEMORY_XP from BaseAddress 0x%llx of length 0x%llx!\n",
```
Comment (kuqin12) [inline ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Ditto.

**Review by kuqin12 — APPROVED** [review]

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c (hunk 1 above)
Comment (os-d) [reply ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Updated.

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c (hunk 2 above)
Comment (os-d) [reply ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Updated.

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c (hunk 3 above)
Comment (os-d) [reply ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Updated.

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c (hunk 4 above)
Comment (os-d) [reply ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Updated.

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c (hunk 5 above)
Comment (os-d) [reply ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Updated.

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c (hunk 6 above)
Comment (os-d) [reply ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Updated.

File: ArmPkg/Drivers/CpuDxe/MemoryAttribute.c (hunk 7 above)
Comment (os-d) [reply ArmPkg/Drivers/CpuDxe/MemoryAttribute.c]: Updated.

**Review by Flickdm — APPROVED** [review]

**Review by apop5 — APPROVED** [review]

**Review by kenlautner — APPROVED** [review]

**Pull request merged**